File: clients/go/consumer.go
// pgque-go -- Go client for PgQue
// Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

package pgque

import (
	"context"
	"fmt"
	"log"
	"time"
)

// HandlerFunc processes a single message. Returning a non-nil error
// causes the Consumer to issue a per-message Nack for that message
// (routing it to retry_queue or, once max_retries is exceeded, to the
// dead_letter table). Other messages in the same batch are still
// dispatched to their own handlers; the batch as a whole is Ack'd
// only if every per-message Nack call returned without error. A
// panic raised inside the handler is recovered and routed identically
// to a returned error.
type HandlerFunc func(ctx context.Context, msg Message) error

// consumerBackend is the subset of Client used by Consumer. Defining
// it as an interface keeps the Consumer testable: a stub backend can
// simulate Nack failures without a live database.
type consumerBackend interface {
	Receive(ctx context.Context, queue, consumer string, maxMessages int) ([]Message, error)
	Ack(ctx context.Context, batchID int64) error
	Nack(ctx context.Context, batchID int64, msg Message, opts ...NackOption) error
}

// Consumer polls a queue and dispatches messages to registered
// handlers. Create one via Client.NewConsumer.
type Consumer struct {
	backend       consumerBackend
	queue         string
	name          string
	pollInterval  time.Duration
	maxMessages   int
	handlers      map[string]HandlerFunc
	unknownPolicy UnknownHandlerPolicy
}

// Handle registers fn as the handler for messages whose Type matches
// eventType. Messages with no registered handler are dispatched
// according to the Consumer's UnknownHandlerPolicy: by default
// (NackUnknown) each is logged and Nack'd individually, routing it to
// retry_queue or eventually the DLQ; with
// WithUnknownHandlerPolicy(AckUnknown) each is logged and silently
// skipped (the surrounding batch is still Ack'd).
func (c *Consumer) Handle(eventType string, fn HandlerFunc) {
	c.handlers[eventType] = fn
}

// dispatchWithRecover calls fn and converts any panic into a non-nil
// error so that the caller can nack the message and keep polling.
func (c *Consumer) dispatchWithRecover(ctx context.Context, fn HandlerFunc, msg Message) (retErr error) {
	defer func() {
		if r := recover(); r != nil {
			retErr = fmt.Errorf("handler panic: %v", r)
		}
	}()
	return fn(ctx, msg)
}

// Start begins the poll loop and blocks until ctx is cancelled. On
// Receive errors it logs and retries after the configured poll
// interval.
//
// Per-batch dispatch semantics (the runtime contract of this loop):
//
//   - Each message is delivered to its registered handler. A handler
//     that returns a non-nil error, or panics, has its message
//     individually Nack'd (routed to retry_queue, eventually the DLQ).
//     The handler error is logged and dispatch continues to the next
//     message in the batch.
//   - Messages with no registered handler follow the configured
//     UnknownHandlerPolicy: NackUnknown (default) logs and Nacks each;
//     AckUnknown logs and skips each, leaving the batch Ack to handle
//     them.
//   - After every message has been processed, the batch is Ack'd —
//     unless one of the per-message Nack calls returned an error, in
//     which case the batch is left unacked so PgQue redelivers it on
//     the next Receive. Acking a batch whose Nack failed would silently
//     drop the failure information.
func (c *Consumer) Start(ctx context.Context) error {
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		msgs, err := c.backend.Receive(ctx, c.queue, c.name, c.maxMessages)
		if err != nil {
			log.Printf("pgque: receive error: %v", err)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(c.pollInterval):
			}
			continue
		}

		if len(msgs) == 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(c.pollInterval):
			}
			continue
		}

		var batchID int64
		nackFailed := false
		for _, msg := range msgs {
			batchID = msg.BatchID
			handler, ok := c.handlers[msg.Type]
			if !ok {
				if c.unknownPolicy == AckUnknown {
					log.Printf("pgque: no handler registered for event type %q, skipping message %d (AckUnknown policy)", msg.Type, msg.MsgID)
					continue
				}
				log.Printf("pgque: no handler registered for event type %q, nacking message %d", msg.Type, msg.MsgID)
				if nackErr := c.backend.Nack(ctx, batchID, msg); nackErr != nil {
					log.Printf("pgque: nack error for unhandled type %s: %v", msg.Type, nackErr)
					nackFailed = true
				}
				continue
			}
			if handlerErr := c.dispatchWithRecover(ctx, handler, msg); handlerErr != nil {
				log.Printf("pgque: handler error for %s: %v", msg.Type, handlerErr)
				if nackErr := c.backend.Nack(ctx, batchID, msg); nackErr != nil {
					log.Printf("pgque: nack error for %s: %v", msg.Type, nackErr)
					nackFailed = true
				}
				continue
			}
		}

		if batchID != 0 {
			if nackFailed {
				log.Printf("pgque: skipping ack for batch %d due to prior nack failures; PgQue will redeliver", batchID)
				continue
			}
			if err := c.backend.Ack(ctx, batchID); err != nil {
				log.Printf("pgque: ack error: %v", err)
			}
		}
	}
}

File: clients/typescript/src/consumer.ts
// pgque -- TypeScript client for PgQue
// Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

import type { Client } from './client.js';
import type { ConsumerOptions, HandlerFunc, Message } from './types.js';

const DEFAULT_MAX_MESSAGES = 2_147_483_647; // PostgreSQL int4 max; request the whole batch by default.

// Reconnect backoff steps (ms) for the LISTEN connection on disconnect.
// Capped at the consumer's pollInterval.
const LISTEN_BACKOFF_MS = [1_000, 2_000, 5_000];

/**
 * High-level consumer that polls `pgque.receive`, dispatches each message
 * to a per-event-type handler, and finalizes the batch with `ack` (or
 * per-message `nack` on handler failure / unknown event type).
 *
 * **LISTEN/NOTIFY wakeup:** the consumer opens a dedicated `pg.Client`
 * connection and issues `LISTEN pgque_<queue>`. When the PgQue ticker
 * emits a `pg_notify`, the consumer wakes immediately instead of waiting
 * for the next poll interval. Polling is retained as a safety net for
 * missed notifications and network drops.
 *
 * Usage:
 * ```ts
 * const consumer = client.newConsumer('orders', 'order_worker');
 * consumer.handle('order.created', async (msg) => { ... });
 *
 * const ac = new AbortController();
 * await consumer.start(ac.signal);
 * ```
 */
export class Consumer {
  private readonly handlers = new Map<string, HandlerFunc>();
  private readonly pollIntervalMs: number;
  private readonly maxMessages: number;
  private readonly unknownHandlerPolicy: 'ack' | 'nack';
  private readonly logger: Pick<Console, 'warn' | 'error'>;
  private readonly listenClientFactory: ConsumerOptions['_listenClientFactory'];

  /** @internal — use {@link Client.newConsumer}. */
  constructor(
    private readonly client: Client,
    private readonly queue: string,
    private readonly name: string,
    opts: ConsumerOptions = {},
  ) {
    this.pollIntervalMs = opts.pollInterval ?? 30_000;
    this.maxMessages = opts.maxMessages ?? DEFAULT_MAX_MESSAGES;
    this.unknownHandlerPolicy = opts.unknownHandlerPolicy ?? 'nack';
    this.logger = opts.logger ?? console;
    this.listenClientFactory = opts._listenClientFactory;
  }

  /** Register a handler for `eventType`. Replaces any previous handler. */
  handle(eventType: string, fn: HandlerFunc): void {
    this.handlers.set(eventType, fn);
  }

  /**
   * Start the poll loop. Resolves when `signal` is aborted; rejects only
   * on terminal errors that should bubble up (the routine `Receive`/`Ack`
   * errors are logged and the loop continues).
   *
   * **Abort granularity:** aborting the signal interrupts the inter-poll
   * `sleep()` immediately, but does **not** cancel an in-flight
   * `client.receive()` call. If a `receive()` round-trip is in progress
   * when the signal fires, the loop will drain that call to completion
   * before exiting.
   *
   * **LISTEN/NOTIFY:** a dedicated `pg.Client` opens `LISTEN pgque_<queue>`
   * so the loop wakes as soon as the ticker fires rather than waiting for the
   * full `pollInterval`. Polling remains active as a fallback safety net.
   */
  async start(signal?: AbortSignal): Promise<void> {
    // notifyResolve is the resolve function for the current "wait for notify"
    // promise. Calling it wakes the poll loop early.
    let notifyResolve: (() => void) | null = null;

    const makeNotifyPromise = (): Promise<void> =>
      new Promise<void>((resolve) => {
        notifyResolve = resolve;
      });

    let currentNotifyPromise = makeNotifyPromise();

    const onNotification = (): void => {
      const resolve = notifyResolve;
      if (resolve) {
        notifyResolve = null;
        resolve();
        // Pre-arm next cycle so the handler is always ready.
        currentNotifyPromise = makeNotifyPromise();
      }
    };

    // Spin up LISTEN connection in the background; reconnect on drops.
    const channel = `pgque_${this.queue}`;

    const connectListen = async (): Promise<void> => {
      if (!this.listenClientFactory) {
        // No factory injected — skip LISTEN; poll-only fallback.
        return;
      }
      let backoffIdx = 0;
      while (!signal?.aborted) {
        try {
          const client = await this.listenClientFactory();

          client.on('notification', onNotification);
          client.on('error', (_err: Error) => {
            // Suppress the unhandled rejection that Node.js emits for
            // EventEmitter 'error' events; the disconnect is handled below.
          });

          await client.connect();
          await client.query(`LISTEN ${quoteIdentifier(channel)}`);
          backoffIdx = 0; // reset on successful connect

          // Block until aborted or the connection emits 'end'.
          if (!signal?.aborted) {
            await new Promise<void>((resolve) => {
              const onEnd = (): void => resolve();
              const onAbort = (): void => {
                client.removeListener('end', onEnd);
                resolve();
              };
              client.once('end', onEnd);
              signal?.addEventListener('abort', onAbort, { once: true });
            });
          }

          // Clean up: UNLISTEN + end.
          client.removeListener('notification', onNotification);
          try {
            await client.query(`UNLISTEN ${quoteIdentifier(channel)}`);
          } catch {
            // ignore: connection may already be broken
          }
          try {
            await client.end();
          } catch {
            // ignore
          }

          if (signal?.aborted) return;

          // Unexpected disconnect — fall through to reconnect logic below.
          this.logger.warn(`pgque: LISTEN connection dropped for ${channel}, reconnecting`);
        } catch (err) {
          this.logger.error(
            `pgque: LISTEN connect error for ${channel}: ${formatErr(err)}, reconnecting`,
          );
        }

        if (signal?.aborted) return;

        // Exponential backoff capped at pollInterval.
        const delay = Math.min(
          LISTEN_BACKOFF_MS[backoffIdx] ?? LISTEN_BACKOFF_MS[LISTEN_BACKOFF_MS.length - 1]!,
          this.pollIntervalMs,
        );
        backoffIdx = Math.min(backoffIdx + 1, LISTEN_BACKOFF_MS.length - 1);
        await sleep(delay, signal);
      }
    };

    // Launch the LISTEN loop; don't await — it runs alongside the poll loop.
    const listenDone = connectListen();

    // Poll loop.
    while (!signal?.aborted) {
      let msgs: Message[];
      try {
        msgs = await this.client.receive(this.queue, this.name, this.maxMessages);
      } catch (err) {
        this.logger.error(`pgque: receive error: ${formatErr(err)}`);
        await Promise.race([sleep(this.pollIntervalMs, signal), currentNotifyPromise]);
        currentNotifyPromise = makeNotifyPromise();
        continue;
      }

      if (msgs.length === 0) {
        await Promise.race([sleep(this.pollIntervalMs, signal), currentNotifyPromise]);
        currentNotifyPromise = makeNotifyPromise();
        continue;
      }

      let batchId: bigint | null = null;
      let anyNackFailed = false;
      for (const msg of msgs) {
        batchId = msg.batchId;
        const handler = this.handlers.get(msg.type);
        if (!handler) {
          if (this.unknownHandlerPolicy === 'ack') {
            this.logger.warn(
              `pgque: no handler registered for event type "${msg.type}", acking msg ${msg.msgId} (unknownHandlerPolicy='ack')`,
            );
            // Fall through; the batch ack at the end of the loop covers it.
            continue;
          }
          this.logger.warn(
            `pgque: no handler registered for event type "${msg.type}", nacking msg ${msg.msgId}`,
          );
          if (!(await this.tryNack(batchId, msg, 'unknown event type'))) {
            anyNackFailed = true;
          }
          continue;
        }
        try {
          await handler(msg);
        } catch (err) {
          this.logger.error(`pgque: handler error for "${msg.type}": ${formatErr(err)}`);
          if (!(await this.tryNack(batchId, msg, 'handler error'))) {
            anyNackFailed = true;
          }
        }
      }

      if (batchId !== null) {
        if (anyNackFailed) {
          // At least one required nack failed. Skip ack so PgQ redelivers
          // the batch on the next poll instead of advancing the consumer
          // past messages we couldn't route.
          this.logger.error(
            `pgque: skipping ack for batch ${batchId}; one or more nacks failed and the batch will be redelivered`,
          );
        } else {
          try {
            await this.client.ack(batchId);
          } catch (err) {
            this.logger.error(`pgque: ack error: ${formatErr(err)}`);
          }
        }
      }
    }

    // Signal aborted: wake the notify promise so the listen loop exits cleanly.
    onNotification();
    // Ensure the LISTEN loop also exits (it checks signal?.aborted).
    await listenDone;
  }

  /** Returns true if the nack succeeded, false if it threw (and was logged). */
  private async tryNack(batchId: bigint, msg: Message, reason: string): Promise<boolean> {
    try {
      await this.client.nack(batchId, msg, { reason });
      return true;
    } catch (err) {
      this.logger.error(`pgque: nack error for "${msg.type}": ${formatErr(err)}`);
      return false;
    }
  }
}

function formatErr(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) {
      resolve();
      return;
    }
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve();
    }, ms);
    const onAbort = (): void => {
      clearTimeout(timer);
      resolve();
    };
    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

/**
 * Quote a PostgreSQL identifier (channel name) for safe use in LISTEN/UNLISTEN.
 * Doubles any double-quote characters and wraps in double quotes.
 */
function quoteIdentifier(name: string): string {
  return `"${name.replace(/"/g, '""')}"`;
}
